pkg/agent/secondarynetwork/podwatch/controller.go

```diff
@@ -0,0 +1,376 @@
+        if err != nil {
+               return fmt.Errorf("GetPodContainerDeviceIDs failed: %v", err)
+        }
+
+        if err = pc.cniServer.GetPodConfigurator().ConfigureSriovSecondaryInterface(
+                                      containerInfo.PodName,
+                                      containerInfo.PodNameSpace,
+                                      containerInfo.ContainerID,
+                                      containerInfo.ContainerNetNS,
+                                      netinfo.InterfaceName,
+                                      containerInfo.MTU,
+                                      podSriovVFDeviceID[0],
+                                      result,
+                                      ); err != nil {
+                 klog.ErrorS(err, "Failed to configure interfaces for container ", containerInfo.ContainerID, err)
+                 return err
+        }
+	return nil
+}
+
+// Configure Secondary Network Interface.
+func (pc *PodController) configureSecondaryInterface(pod *corev1.Pod, netinfo *SecondaryNetworkObject, podCNIInfo []*cnipodcache.CNIConfigInfo, cniconfig []byte) error {
+	var err error
+        var ipamResult *current.Result
+        var cmdArgs *invoke.Args
+
+        for _, containerInfo := range podCNIInfo {
+	       if netinfo.InterfaceType == sriovInterfaceType {
+                     cmdArgs = &invoke.Args{Command: string("ADD"), ContainerID: containerInfo.ContainerID,
+                             NetNS: containerInfo.ContainerNetNS, IfName: containerInfo.ContainerIfDev,
+                             Path: cniPath}
+
+                     ipamResult, err = ipam.GetIPAMSubnetAddress(cniconfig, cmdArgs)
+                     if err != nil {
+                            return errors.New("secondary network IPAM failed")
+                     }
+
+                     result := &current.Result{CNIVersion: containerInfo.CNIVersion}
+                     result.IPs = ipamResult.IPs
+                     result.Routes = ipamResult.Routes
+	             // Set result.Interface to container interface.
+                     for _, ip := range result.IPs {
+                           ip.Interface = current.Int(1)
+                     }
+                     // If secondary interface name not provided in Pod annotation, assign default name.
+                     if len(netinfo.InterfaceName) == 0 {
+                           netinfo.InterfaceName = netDev
+                     }
+		     // Configure SRIOV as a secondary network interface
+		     if err = pc.configureSriovAsSecondaryInterface(pod, netinfo, containerInfo, result); err != nil {
+			  // SRIOV interface creation failed. Free allocated IP address
+                          ipam.DelIPAMSubnetAddress(cniconfig, cmdArgs)
+			  return err
+                     }
+               } else {
+                     klog.ErrorS(err, "InterfaceType not supported for container ", containerInfo.ContainerID)
```

Comment:
Probably Pod name is better than container ID in the log message. But up to you.